cpp/daal/include/services/internal/sycl/math/mkl_lapack.h

```diff
@@ -106,50 +125,72 @@ struct MKLPotrs
     MKLPotrs(cl::sycl::queue & queue) : _queue(queue) {}
 
     Status operator()(const math::UpLo uplo, const size_t n, const size_t ny, Buffer<algorithmFPType> & a, const size_t lda,
-                      Buffer<algorithmFPType> & b, const size_t ldb, cl::sycl::buffer<algorithmFPType, 1> & scratchpad)
+                      Buffer<algorithmFPType> & b, const size_t ldb)
     {
-        Status status;
-        const ::oneapi::fpk::uplo uplomkl = uplo == math::UpLo::Upper ? ::oneapi::fpk::uplo::upper : ::oneapi::fpk::uplo::lower;
+        const ::oneapi::fpk::uplo uplomkl        = uplo == math::UpLo::Upper ? ::oneapi::fpk::uplo::upper : ::oneapi::fpk::uplo::lower;
+        const std::int64_t minimalScratchpadSize = ::oneapi::fpk::lapack::potrs_scratchpad_size<algorithmFPType>(_queue, uplomkl, n, ny, lda, ldb);
+        return this->operator()(uplo, n, ny, a, lda, b, ldb, minimalScratchpadSize);
+    }
 
-        cl::sycl::buffer<algorithmFPType, 1> a_sycl_buff = a.toSycl(status);
-        DAAL_CHECK_STATUS_VAR(status);
+private:
+    Status operator()(const math::UpLo uplo, const size_t n, const size_t ny, Buffer<algorithmFPType> & a, const size_t lda,
+                      Buffer<algorithmFPType> & b, const size_t ldb, const std::int64_t scratchpadSize)
+    {
+        using namespace daal::services;
 
-        cl::sycl::buffer<algorithmFPType, 1> b_sycl_buff = b.toSycl(status);
-        DAAL_CHECK_STATUS_VAR(status);
+        services::Status status;
+        const ::oneapi::fpk::uplo uplomkl = uplo == math::UpLo::Upper ? ::oneapi::fpk::uplo::upper : ::oneapi::fpk::uplo::lower;
 
+#ifdef DAAL_SYCL_INTERFACE_USM
+        if (a.isUSMBacked())
         {
-            const size_t minimalScratchpadSize =
-                size_t(::oneapi::fpk::lapack::potrs_scratchpad_size<algorithmFPType>(_queue, uplomkl, n, ny, lda, ldb));
-            if (scratchpad.get_count() < minimalScratchpadSize) return ErrorMemoryAllocationFailed;
+            auto a_usm = a.toUSM(status);
+            DAAL_CHECK_STATUS_VAR(status);
+
+            auto b_usm = b.toUSM(status);
+            DAAL_CHECK_STATUS_VAR(status);
+
+            algorithmFPType * scratchpad = nullptr;
+            if (scratchpadSize > 0)
+            {
+                scratchpad = cl::sycl::malloc<algorithmFPType>(scratchpadSize, _queue, cl::sycl::usm::alloc::shared);
+                if (scratchpad == nullptr) return ErrorMemoryAllocationFailed;
+            }
+
+            status |= catchSyclExceptions([&]() mutable {
+                ::oneapi::fpk::lapack::potrs(_queue, uplomkl, n, ny, a_usm.get(), lda, b_usm.get(), ldb, scratchpad, scratchpadSize);
+                _queue.wait_and_throw();
+            });
+
+            if (scratchpadSize > 0) cl::sycl::free(scratchpad, _queue);
+
+            scratchpad = nullptr;
```

Comment:
Shouldn't we just return status?
```suggestion
            scratchpad = nullptr;
            return Status();
```

Reply:
Empty status means "all is ok", but here might be some errors in fact. Correct status is returned once at the end of this function.